DOCS-8072-4.3-tuning-backpressure-maxconcurrency-kt (#1411)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/tuning-backpressure-maxconcurrency.adoc

> = Back-Pressure and Max Concurrency

Reviewer: ```suggestion
= Back Pressure Avoidance Using the maxConcurrency Parameter
```
OR
```suggestion
= Back Pressure Avoidance Using Maximum Concurrency
```

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Under heavy load, Mule runtime engine (Mule) might not have resources available to handle a specific event. This issue might occur because all threads are busy and the current flow’s concurrency has been exceeded already. +

Reviewer: ```suggestion
Under heavy load, Mule runtime engine (Mule) might not immediately have the resources available to process a specific event. This issue might occur because all threads are busy and the current flow’s concurrency limit is reached. When this occurs, Mule logs the following message: +
```

> In case Mule cannot handle an event, Mule logs a message about the condition: +

Reviewer: omit per above
```suggestion
```

`Flow 'flowName' is unable to accept new events at this time` +

> Also known as back-pressure. Mule back-pressure actions are specific to each connector’s source.

Reviewer: ```suggestion
This condition is called _back pressure_, and Mule responds in a manner that is specific to each connector’s source.
```

Reviewer: I recommend moving from this sentence above directly (in the same paragraph) to the text beginning with "Although Mule applies back-pressure on its own,..."
In fact, consider reworking the above text to
```suggestion
This condition is called _back pressure_, and although Mule is designed to lessen back pressure, you can enhance this capability by increasing the value of the `maxConcurrency` parameter.
```
It seems to me that all the other text, about connector source, producer components, and consumer components is distracting and unnecessary, and that we should document only the use of the parameter to solve back pressure issues.

Reviewer: Nevertheless, copyedited below.

> When there is a producer component that outputs information at a higher rate than a consumer component can handle, you can use the flow’s `maxConcurrency` parameter to set the maximum number of concurrent messages that a flow can process at any given time. +

Reviewer: ```suggestion
When a producer component outputs information more quickly than a consumer component can process it, you can increase the value of the consumer flow `maxConcurrency` parameter to improve processing capacity, whether the components are internal or external, flows or sources, and so on.+
```

> This scenario can manifest in various forms, as those components could be internal/external, flows/sources, etc.

Reviewer: Omit per above suggestion.
```suggestion
```

> Although Mule applies back-pressure on its own, the `maxConcurrency` parameter helps to keep under control the performance of your flows, especially when a backend server may affect when having high concurrency. See how to set the parameter in the following example:

Reviewer: See previous suggestions.

Do we need to mention and, if so, explain how Mule applies solutions to back pressure (applies SOLUTIONS, right, not applies the back pressure itself?)

```suggestion
Although Mule automatically applies internal methods to control back pressure, you can customize the `maxConcurrency` parameter to help control flow performance of your flows:
```
Again, I think that this is wrong: Although Mule **applies back-pressure** on its own,
and this does not make sense: "especially when a backend server may affect when having high concurrency"
Does this mean that a backend server can cause high concurrency or increase existing high concurrency?

[source,xml,linenums]
----
<flow name="perf-test-flow" maxConcurrency="${flow.max.concurrency}">
----

> Similarly, you can limit the amount of records that can be processed in a batch job setting the `maxConcurrency` parameter like the following example:

Reviewer: ```suggestion
Similarly, you can limit the number of records that can be processed in a batch job by setting the `maxConcurrency` parameter as in the following example:
```
